[DI - DI 프레임워크 구현] - 4단계 - @Configuration 설정(힌트)

Wu22e opened the pull request:
안녕하세요 용대님!
4단계 미션 PR 요청 드립니다!

피드백 주신 ApplicationContext 에 대한 통합테스트(수동, 자동 스캔 빈 주입 통합)를 위한 테스트를 추가해 보았습니다.

또, BeanDefinition 을 인터페이스로 변경해서 클래스 빈, 메서드 빈에 대한 책임을 분리해 볼 수 있었습니다.

혹시 잘못된 부분이 있다면 피드백 주시면 감사하겠습니다!

이번 미션 리뷰도 잘 부탁드립니다!

Wu22e committed:
test : BeanFactoryTest register, scanner 이용하여 setup 하도록 변경

Wu22e committed:
test : ApplicationContext 를 이용한 자동, 수동 스캔 통합 테스트

Wu22e committed:
docs : 기능 요구사항 수정 (BeanDefinition 인터페이스로 변경)

Wu22e committed:
docs : 기능 목록 수정 (BeanDefinition 인터페이스로 변경)

Wu22e committed:
refactor : BeanDefinition 인터페이스로 변경

- ClassBeanDefinition, MethodBeanDefinition 구현체 추가

Wu22e committed:
Merge remote-tracking branch 'origin/step4' into step4

yongdae commented on README.md:
```diff
@@ -31,10 +31,12 @@
-  - 빈에 대한 클래스 정보와 메서드 정보를 필드로 가진다.
-  - 해당 빈이 클래스 타입 빈일 경우, 클래스의 타입을 필드로 가지고, 메서드 정보는 null 이다.
-  - 해당 빈이 메서드 타입 빈일 경우, 클래스 타입과 메서드 정보를 모두 가진다.
+- BeanDefinition 인터페이스
```
👍

yongdae commented on src/test/java/core/di/factory/ApplicationContextTest.java:
```diff
@@ -0,0 +1,59 @@
+
+import static org.junit.jupiter.api.Assertions.*;
+
+class ApplicationContextTest {
```
ApplicationContextTest는 ApplicationContext에 대한 테스트라기보다는
BeanFactory에 대한 테스트 인 것 같아요.

    ApplicationContext applicationContext = new ApplicationContext(IntegrationConfig.class);

이런식으로 IntegrationConfig를 사용해서 ApplicationContext 테스트를 만들어보면 어떨까요?

Wu22e replied:
Application 내부에서 init 하는 부분을 BeanFactory 하듯이 하고 있었군요 .. 수정하도록 하겠습니다!

yongdae commented on src/main/java/core/di/BeanDefinitionRegistry.java:
```diff
@@ -19,7 +19,7 @@ public BeanDefinitionRegistry() {
     public void registerClassPathBeans(Set<Class<?>> classPathBeanClasses) {
         for (Class<?> classPathBeanClass : classPathBeanClasses) {
-            this.classBeanDefinitions.add(new BeanDefinition(classPathBeanClass));
+            this.classBeanDefinitions.add(new ClassBeanDefinition(classPathBeanClass));
```
DI를 활용해서 BeanDefinitionRegistry가 ClassBeanDefinition, MethodBeanDefinition 아는 대신 BeanDefinition만 알고 사용하는 형태로 바꾸어보면 어떨까요?

Wu22e replied:
꼼꼼한 리뷰 감사합니다! 구현체에 대한 의존도를 없애는게 좋을 것 같네요!

Wu22e replied:
꼼꼼한 리뷰 감사합니다! 구현체에 대한 의존도를 없애는게 좋을 것 같네요!

yongdae commented on src/main/java/core/di/BeanDefinition.java:
```diff
@@ -1,66 +1,10 @@
-                ", beanMethod=" + beanMethod +
-                '}';
-    }
+public interface BeanDefinition {
```
의견드린 부분을 잘 반영해주셨네요. 💯 

BeanDefinition처럼 빈을 검색하는 개념(AnnotatedBeanDefinitionReader, ClassPathBeanDefinitionScanner)도 인터페이스를 이용해 정의하고 구현체를 만들어보면 어떨까요? 🤔

yongdae commented on src/test/java/core/di/ClassBeanDefinitionTest.java:
```diff
@@ -0,0 +1,27 @@
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+import static org.junit.jupiter.api.Assertions.*;
+
+class ClassBeanDefinitionTest {
```
ClassBeanDefinition, MethodBeanDefinition을 구현체를 만드시고 검증까지 잘 만들어주셨네요. 💯

yongdae reviewed (changes requested):
이전 단계에서 스프링을 잘 참고해서 만들어주신 덕분에
크게 힌트를 적용할 부분이 없이 잘 설계를 해주셨던 것 같아요!

진행해주신 내용에 대해 몇 가지 의견을 남겨놓았습니다~
확인하시고 다시 리뷰 요청 부탁드릴께요. 🔥🔥

Wu22e committed:
refactor : ApplicationContext 통합 테스트 수정

- ApplicationContext 의 initialize 를 바로 사용하도록 수정
- Config 클래스의 ComponentScan 에 value 값이 지정되지 않을 경우 해당 ComponentScan 클래스 패키지 부터 스캔하도록 하는 로직 추가

Wu22e committed:
refactor : BeanScanner 인터페이스 구현

- BeanDefinitionRegistry 에서 BeanDefinition 구현체 의존 제거 (의존성 주입을 통해 해결)
- scan, register 메서드 -> scan 메서드로 통합

Wu22e commented:
안녕하세요 용대님! 꼼꼼한 피드백 감사합니다!
의견 주신 부분들을 반영해 보았습니다.

추가적으로 BeanScanner 인터페이스 추출에서 좀 고민이 들었는데, 우선 지금 상황은 scan, register 메서드를 scan 하나의 메서드로 통합한 상황입니다. 

여기서 제가 생각했던 것은 ClassPathBeanDefinitionScanner 입장에서 scan 메서드의 파라미터가 Object... basePackages 가 더 직관적으로 이해가 잘 간다고 생각해서 이전 구조를 그대로 사용하는것도 크게 문제되지는 않아 보였습니다.

이전 구조가 좋을지 아니면 지금 바뀐 구조인 scan 메서드의 통합이 좋을까요?

아니면 공통 메서드로 가져가지 않고
```java
public interface BeanScanner {
	void scan(Object… basePackages);
	void register(Set<Class<?> configurationClasses);
}

public class AnnotatedBeanDefinitionReader implements BeanScanner {
	@Overrid
	void scan(Object… basePackages) {
		// 예외 발생
	}

	@Overrid
	void register(Set<Class<?> configurationClasses) {
		// 스캔 & 등록
	}
}




public class ClassPathBeanDefinitionScanner implements BeanScanner {
	@Overrid
	void scan(Object… basePackages) {
		// 스캔 & 등록
	}

	@Overrid
	void register(Set<Class<?> configurationClasses) {
		// 예외 발생
	}
}

```
다음과 같이 인터페이스로 나눌 수 있겠다는 생각도 했는데

 어떤 방식이 현 구조에서 좋은 구조일지 용대님의 생각이 궁금합니다!

yongdae commented on src/main/java/core/di/BeanScanner.java:
```diff
@@ -0,0 +1,7 @@
+
+import java.util.Set;
+
+public interface BeanScanner {
```
인터페이스로 분리한 것에대해 한 가지 [질문](https://github.com/next-step/jwp-di/pull/150#issuecomment-1257226561)을 주셨는데요.

주관적인 판단으로는 지금과 같이 하나의 메소드만 가지는게 좋다고 생각합니다.
scan(), register()를 의견 주신 것과 같은 형태로 사용한다면 인터페이스로 정의하는 것은 불편함만을 가져다 줄 뿐 입니다.

AnnotatedBeanDefinitionReader, ClassPathBeanDefinitionScanner가 아닌 또 다른 형태로 매개변수를 받아 빈을 검색해야하는 구현체가 필요하게 된다면 scan(), register() 가 아닌 또 다른 메소드를 정의해야하는데 이는 인터페이스를 사용하는 추상화를 벗어난 방식이라고 생각해요. 더군다나 예외를 던지는 부분이 늘어나게 되는데 이는 좋은 방법이라고 생각되지는 않아요. 😭

yongdae commented on src/main/java/core/di/factory/ApplicationContext.java:
```diff
@@ -23,24 +23,16 @@ public ApplicationContext(Class<?>... configurationClasses) {
-        reader.register(this.configurationClasses);
+        BeanScanner annotatedBeanDefinitionScanner = new AnnotatedBeanDefinitionScanner(beanDefinitionRegistry);
+        annotatedBeanDefinitionScanner.scan(this.configurationClasses);
 
-        ClassPathBeanDefinitionScanner scanner = new ClassPathBeanDefinitionScanner(beanDefinitionRegistry);
-        scanner.scan(getBasePackages(this.configurationClasses));
+        BeanScanner classPathBeanDefinitionScanner = new ClassPathBeanDefinitionScanner(beanDefinitionRegistry);
+        classPathBeanDefinitionScanner.scan(this.configurationClasses);
```
BeanScanner를 리스트로 정의해서 사용하는 형태를 해보았어도 재미있었을 것 같아요. :)

    List<BeanScanner> beanScanners = new ArrayList<>();

yongdae commented on src/main/java/core/di/BeanDefinitionRegistry.java:
```diff
@@ -1,39 +1,25 @@
-            registerConfigurationBean(configurationClass);
-        }
-    }
-
-    private void registerConfigurationBean(Class<?> configurationClass) {
-        Set<Method> beanMethods = ReflectionUtils.getAllMethods(configurationClass, ReflectionUtilsPredicates.withAnnotation(Bean.class));
-
-        for (Method method : beanMethods) {
-            this.methodBeanDefinitions.add(new BeanDefinition(configurationClass, method));
+    public void registerConfigurationBeans(Set<BeanDefinition> methodBeanDefinitions) {
+        for (BeanDefinition methodBeanDefinition : methodBeanDefinitions) {
+            this.methodBeanDefinitions.add(methodBeanDefinition);
```
피드백을 잘 반영해주셨네요. 💯 

BeanDefinitionRegistry도 BeanDefinitions처럼 인터페이스로 정의하면
필요에 따라 registerClassPathBeans(), registerConfigurationBeans()를 나누어 사용하지 않게 할 수 있었을 것 같아요. 😄

yongdae reviewed (approved):
의견 드린 부분을 잘 고민해서 반영해주셨네요. 👍 
그럼 이만 이번 미션을 마무리 하도록 하겠습니다.

의견 주신 부분은 리뷰를 통해 답변드렸어요.
덧붙여 의견드린 부분과 같이 확인 부탁드릴께요~

그럼 마지막까지 힘내주세요!! 수고 많으셨습니다. 🙇

Wu22e commented:
안녕하세요 용대님!
이번에 남겨주신 리뷰들은 개인 브랜치에 생각해보고 반영해두도록 하겠습니다!
이번 미션 진행하면서 꼼꼼하게 놓칠 부분들에 대한 피드백과 생각해볼만한 리뷰 남겨주셔서 감사했습니다!

남은 미션까지 잘 끝내보도록 하겠습니다~!
